### PR TITLE
If there is no replica based on the partition map, fall back to Cassandra behavior

### PR DESCRIPTION
**Pull request by tedyu**
See https://yugabyte-db.slack.com/archives/CG0KQF0GG/p1600685168035400 for background.

The change has been verified by Paramiyoti.

### CODE REVIEW

**Comment by tedyu** [comment]
ping @m-iancu for review.

**Comment by tedyu** [comment]
ping @m-iancu

**Comment by tedyu** [comment]
ping @m-iancu for review.

**Comment by tedyu** [comment]
ping @m-iancu for review.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala [hunk 1]
```diff
@@ -79,12 +79,18 @@ private[connector] class CassandraPartitionGenerator[V, T <: Token[V]](
...
+      var ranges = partitionMap.values.toSeq.map { partition =>
         val startToken = tokenFactory.tokenFromString(PartitionAwarePolicy.YBToCqlHashCode(partition.getStartKey()).toString)
         val endToken = tokenFactory.tokenFromString(PartitionAwarePolicy.YBToCqlHashCode(partition.getEndKey()).toString)
         val hosts = partition.getHosts().map(_.getAddress).toSet
         new TokenRange(startToken, endToken, hosts, tokenFactory)
       }
+      ranges match {
+        case head #:: rest =>
+          if (head.replicas.isEmpty)
+            ranges = describeRing
```
Comment (m-iancu) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala:92]: why do we only need to check the head here, what about the other entries?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala [hunk 2]
```diff
@@ -79,12 +79,18 @@ private[connector] class CassandraPartitionGenerator[V, T <: Token[V]](
...
       // Compute map from (start) token to corresponding hosts based on the partition map.
       // Converting token map into Sequence of TokenRanges describing the ring for this table.
-      partitionMap.values.toSeq.map { partition =>
+      var ranges = partitionMap.values.toSeq.map { partition =>
         val startToken = tokenFactory.tokenFromString(PartitionAwarePolicy.YBToCqlHashCode(partition.getStartKey()).toString)
         val endToken = tokenFactory.tokenFromString(PartitionAwarePolicy.YBToCqlHashCode(partition.getEndKey()).toString)
         val hosts = partition.getHosts().map(_.getAddress).toSet
         new TokenRange(startToken, endToken, hosts, tokenFactory)
       }
+      ranges match {
```
Comment (m-iancu) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala:88]: Let's add a comment here explaining what is happening.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala [hunk 3]
```diff
@@ -79,12 +79,18 @@ private[connector] class CassandraPartitionGenerator[V, T <: Token[V]](
     val tokenRanges : Seq[TokenRange] = if (partitionMap != null) {
       // Compute map from (start) token to corresponding hosts based on the partition map.
       // Converting token map into Sequence of TokenRanges describing the ring for this table.
-      partitionMap.values.toSeq.map { partition =>
+      var ranges = partitionMap.values.toSeq.map { partition =>
```
Comment (m-iancu) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala:82]: for readability: what is the type of ranges?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala (hunk 1 above)
Comment (tedyu) [reply spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala:92]: The issue had no replica in partition map. In this case checking the head would be enough - we know whether there is  replica returned or not.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala (hunk 3 above)
Comment (tedyu) [reply spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraPartitionGenerator.scala:82]: The type would be Seq[TokenRange]

**Comment by tedyu** [comment]
@m-iancu 
I have updated the PR.

**Review by m-iancu — APPROVED** [review]
Thanks @tedyu, LGTM!

**Pull request merged**